ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -34,6 +38,118 @@ module Winbind = struct
     debug "init_service To be implemented in CP-36085"
 end
 
+let get_service_name () =
+  (fun __context ->
+    Helpers.get_localhost ~__context |> fun host ->
+    Db.Host.get_external_auth_service_name ~__context ~self:host)
+  |> Server_helpers.exec_with_new_task "retrieving external auth service name"
+
+let query_domain_workgroup domain =
+  let key = "Pre-Win2k Domain" in
+  let err_msg = Printf.sprintf "Failed to lookup domain %s workgroup" domain in
+  try
+    let lines =
+      Helpers.call_script ~log_output:On_failure net_cmd
+        ["ads"; "lookup"; "-S"; domain]
+    in
+    match Xapi_cmd_result.of_output_opt ~sep:':' ~key ~lines with
+    | Some v ->
+        v
+    | None ->
+        raise (Auth_service_error (E_LOOKUP, err_msg))
+  with _ -> raise (Auth_service_error (E_LOOKUP, err_msg))
+
+let config_winbind_damon ~domain ~workgroup =
+  let open Xapi_stdext_unix in
+  let smb_config = "/etc/samba/smb.conf" in
+  let conf_contents =
+    String.concat "\n"
+      [
+        "# autogenerated by xapi"
+      ; "[global]"
+      ; "kerberos method = secrets and keytab"
+      ; Printf.sprintf "realm = %s" domain
+      ; "security = ADS"
+      ; "template shell = /bin/bash"
+      ; "winbind offline logon = Yes"
+      ; "winbind refresh tickets = Yes"
+      ; "winbind enum groups = no"
+      ; "winbind enum users = no"
+      ; Printf.sprintf "workgroup = %s" workgroup
+      ; "idmap config * : range = 3000000-3999999"
+      ; Printf.sprintf "idmap config %s: backend = rid" domain
+      ; Printf.sprintf "idmap config %s: range = 2000000-2999999" domain
+      ; "idmap config * : backend = tdb"
+      ; "" (* Empty line at the end *)
+      ]
+  in
+  let len = String.length conf_contents in
+  Unixext.atomic_write_to_file smb_config 0o0644 (fun fd ->
+      let (_ : int) = Unix.single_write_substring fd conf_contents 0 len in
+      ())
+
+let from_config ~name ~err_msg ~config_params =
+  match List.assoc_opt name config_params with
+  | Some v ->
+      v
+  | _ ->
+      raise (Auth_service_error (E_GENERIC, err_msg))
+
+let assert_hostname_valid () =
+  let hostname =
+    (fun __context ->
+      Helpers.get_localhost ~__context |> fun host ->
+      Db.Host.get_hostname ~__context ~self:host)
+    |> Server_helpers.exec_with_new_task "retrieving hostname"
+  in
+  let all_numbers hostname =
+    Re.Perl.re {|^\d+$|} |> Re.Perl.compile |> fun re ->
+    Re.matches re hostname <> []
+  in
+  if all_numbers hostname then
+    raise
+      (Auth_service_error
+         ( E_GENERIC
+         , Printf.sprintf "hostname '%s' cannot contain only digits." hostname
+         ))
+
+let assert_domain_equal_service_name ~service_name ~config_params =
+  (* For legeacy support, if domain exist in config_params, it must be equal to service_name *)
+  let domain_key = "domain" in
+  match List.assoc_opt domain_key config_params with
+  | Some domain ->
+      if domain <> service_name then
```

Comment:
This could be expressed with `when` and an extra case that makes more explicit what happens if `domain = service_name`. Here we have a pattern and an if without an else.